forte/processors/base/data_augment_processor.py

```diff
@@ -16,20 +16,196 @@
+    we prepend an "And" to the second sentence, when modifying the start index
+    of the second Sentence, it should be pushed left to include the new Token.
+    In this case, the is_begin is True.
+
+    Args:
+        index (int): The index to map.
+        old_spans (SortedList): The spans before replacement. It should be
+            a sorted list in ascending order.
+        new_spans (SortedList): The spans after replacement. It should be
+            a sorted list in ascending order.
+        is_begin (bool): True if the input index is the start index of a span.
+        is_inclusive (bool): True if the span constructed by the aligned
+            index should include inserted spans.
+    Returns:
+        The aligned index.
+
+    If the old spans are [0, 1], [2, 3], [4, 6],
+    the new spans are [0, 4], [5, 7], [8, 11],
+    the input index is 3, and there are no insertions,
+    the algorithm will first locate the last span with
+    a begin index less or equal than the target index,
+    ([2,3]), and find the corresponding span in new spans([5,7]).
+    Then we calculate the delta index(7-3=4) and update our
+    input index(3+4=7). The output then is 7.
+
+    Note that when the input index locates inside the old spans,
+    instead of on the boundary of the spans, we just keep the offset
+    of the input index to the old span it was in unchanged. In the
+    above example, if we change the input index from 3 to 5, the
+    output will become 9, because we locates the input index in the
+    third span [4, 6] and use the same offset 5-4=1 to calculate the
+    output 8+1=9.
+
+    When insertion is considered, there will be spans
+    with the same begin index, for example,
+    [0, 1], [1, 1], [1, 2]. The span [1, 1] indicates an insertion
+    at index 1, because the insertion can be considered as a
+    replacement of an empty input span, with a length of 0.
+    The output will depend on whether to include the inserted span,
+    and whether the input index is a begin or an end index.
+
+    If the old spans are [0, 1], [1, 1], [1, 2],
+    the new spans are [0, 2], [2, 4], [4, 5],
+    the input index is 1, the output will be 2 if both
+    is_inclusive and is_begin are True,
+    because the inserted [1, 1] should be included in the span.
+    If the is_inclusive=True but is_begin=False, the output will be
+    4 because the index is an end index of the span.
+    """
+
+    # Get the max index for binary search.
+    max_index: int = old_spans[-1].end + 1
+
+    # This is the last span that has a start index less than
+    # the input index. The position change of this span determines
+    # the modification we will apply to the input index.
+    last_span_ind: int = bisect_right(
+        old_spans, Span(index, max_index)
+    ) - 1
+
+    # If there is an inserted span, it will
+    # always be the first of those spans with
+    # the same begin index.
```

Comment:
Actually I don't understand this sentence too.

Reply:
For example, given spans [1, 1], [1, 2], The inserted span [1, 1] will be in the front of replaced span [1, 2], because it has the smallest end index. Will add this example to the doc.